torchrl/envs/model_based/common.py

```diff
@@ -0,0 +1,150 @@
+        def __init__(self, world_model, device="cpu", dtype=None, batch_size=None):
+            super(MyEnv, self).__init__(world_model, device=device, dtype=dtype, batch_size=batch_size)
+        def _reset(self):
+            td = TensorDict(
+                {
+                    "hidden_observation": torch.randn(*self.batch_size, 4),
+                    "next_hidden_observation": torch.randn(*self.batch_size, 4),
+                    "action": torch.randn(*self.batch_size, 1),
+                },
+                batch_size=self.batch_size,
+            )
+            return td
+        def _set_seed(self, seed: int) -> int:
+            return seed + 1
+    ```
+    Then, you can use this environment as follows:
+    ```python
+        from torchrl.modules import MLP, WorldModelWrapper
+        import torch.nn as nn
+        world_model = WorldModelWrapper(
+            TensorDictModule(
+                MLP(out_features=4, activation_class=nn.ReLU, activate_last_layer=True, depth=0),
+                in_keys=["hidden_observation", "action"],
+                out_keys=["next_hidden_observation"],
+            ),
+            TensorDictModule(
+                nn.Linear(4, 1),
+                in_keys=["hidden_observation"],
+                out_keys=["reward"],
+            ),
+        )
+        world_model = MyWorldModel()
+        env = MyMBEnv(world_model)
+        td = env.reset()
+        env.rollout(td, max_steps=10)
+    ```
+
+
+    Properties:
+        - observation_spec (CompositeSpec): sampling spec of the observations;
+        - action_spec (TensorSpec): sampling spec of the actions;
+        - reward_spec (TensorSpec): sampling spec of the rewards;
+        - batch_size (torch.Size): number of environments contained in the instance;
+        - device (torch.device): device where the env input and output are expected to live
+        - is_done (torch.Tensor): boolean value(s) indicating if the environment has reached a done state since the
+            last reset
+
+    Args:
+        world_model (nn.Module): model that generates world states and its corresponding rewards;
+        device (torch.device): device where the env input and output are expected to live
+        dtype (torch.dtype): dtype of the env input and output
+        batch_size (torch.Size): number of environments contained in the instance
+
+    Methods:
+        step (TensorDict -> TensorDict): step in the environment
+        reset (TensorDict, optional -> TensorDict): reset the environment
+        set_seed (int -> int): sets the seed of the environment
+        rand_step (TensorDict, optional -> TensorDict): random step given the action spec
+        rollout (Callable, ... -> TensorDict): executes a rollout in the environment with the given policy (or random
+            steps if no policy is provided)
+
+    """
+
+    def __init__(
+        self,
+        world_model: TensorDictModule,
+        params: Optional[List[torch.Tensor]] = None,
```

Comment:
if we want to make things functional, could we do like we do for the losses?

```
def __init__(self, model):
     if not isinstance(model, functorch.FunctionalModuleWithBuffers):
         self.model, self.params, self.buffers = model.make_functional_with_buffers()
     else:
        # no idea... perhaps that will never happen
```

or something similar. 
Here I think it's confusing because it's asking to the user to give a functional module and then not to care about the fact that it is functional...

But maybe you have a more specific use case in mind

Reply:
um, the reason i went into funtionalizing it was to be able to pass ensemble weights for the world model. So what we could do would be to functionalize if we don't pass params and buffers?

Reply:
I don't mind that we store the params and buffers independently, but I think this should happen when instantiating the world model, to hide that away from the user. Not sure it makes sense though.
Basically:
```python
self.world_model, self.world_model_params, self.world_model_buffers = make_functional_with_buffers(world_model)
```